### PR TITLE
MongoDB: Added generics to insert and update

### PR DESCRIPTION
**Pull request by fxlrnrpt**
Source code which provides context for the suggested changes: <<https://gist.github.com/keenondrums/0c47052631f16df3732cc9e766a2e754>>

### CODE REVIEW

**Review by CaselIT — APPROVED** [review]
LGTM :+1:

File: types/mongodb/index.d.ts [hunk 1]
```diff
@@ -625,13 +626,13 @@ export interface Collection<TSchema = Default> {
     /** @deprecated use updateOne, updateMany or bulkWrite */
     update(filter: Object, update: Object, options: ReplaceOneOptions & { multi?: boolean }, callback: MongoCallback<WriteOpResult>): void;
     //http://mongodb.github.io/node-mongodb-native/2.1/api/Collection.html#updateMany
-    updateMany(filter: Object, update: Object, callback: MongoCallback<UpdateWriteOpResult>): void;
-    updateMany(filter: Object, update: Object, options?: { upsert?: boolean; w?: any; wtimeout?: number; j?: boolean; }): Promise<UpdateWriteOpResult>;
-    updateMany(filter: Object, update: Object, options: { upsert?: boolean; w?: any; wtimeout?: number; j?: boolean; }, callback: MongoCallback<UpdateWriteOpResult>): void;
+    updateMany<T = Object>(filter: Object, update: T, callback: MongoCallback<UpdateWriteOpResult>): void;
```
Comment (andy-ms) [inline types/mongodb/index.d.ts]: Not sure I see the value of generics here. For some `T`, it takes a `T`. Unless `T` is used in more than one place, That's not any more better than just taking `Object`.

Comment (keenondrums) [reply types/mongodb/index.d.ts]: @andy-ms I listed an example in the gist above. I short generics here will allow us to validate if we can cast DTO to model. Consider the following example:
```
function ctrl(dto: ITestDTO) {
  const test: ITestModel = dto // we can get rid of this line with generics for insert and update
   mongodb.collection('test').insertOne(dto) // here we want to make sure dto can be cast to ITestModel
}
```

Comment (andy-ms) [reply types/mongodb/index.d.ts]: So the generics allow you to assert a type inside an expression... don't think that is worth changing the type definition. You could always use the identity function for a similar purpose: `function id<T>(value: T): T { return value; }`

Comment (keenondrums) [reply types/mongodb/index.d.ts]: Ok. It's just not gonna be as beautiful. Thanks for your time

**Comment by Dante-101** [comment]
I agree with Andy. Providing an explicit type to the function for one argument doesn't help in this scenario. If the `update` is getting a type parameter, why not `filter`?

If the `Collection` interface itself was parameterized intelligently with type containing query and update operators, then it would be so much better.

**Review by rbuckton — CHANGES_REQUESTED** [review]
@keenondrums, do you have a response to @Dante-101's comment above? It seems `Collection` is already generic via the `TSchema` type parameter. Is `T` related to `TSchema` in some fashion? 

From looking at https://docs.mongodb.com/manual/reference/operator/update/, it seems like you want something like:

```ts
export interface UpdateOperators<TSchema> {
  $set?: Partial<TSchema>;
  /* other update operators like $currentDate, $inc, etc. */
}

export interface Collection<TSchema = Default> {
  // ...
  updateOne(filter: Partial<TSchema>, update: UpdateOperators<TSchema>, callback: MongoCallback<UpdateWriteOpResult>): void;
  // etc.
}
```

However, I think one of the reasons that these methods are typed as `Object` is that updates of arrays and embedded documents are permitted using [dot notation](https://docs.mongodb.com/manual/core/document/#document-dot-notation). This kind of update cannot be expressed using `Partial<T>` (or any other type operator), so you would need to keep the existing overloads as a fallback in those cases.

Please let me know if you intend to make any changes based on this feedback.

**Comment by keenondrums** [comment]
@rbuckton I'd like to make the change, just need some extra time to do it.

**Comment by RyanCavanaugh** [comment]
@keenondrums Please address comments from the code reviewers.

**Comment by mhegazy** [comment]
Thanks for your contribution. This PR has failing CI tests and can not be merged in at the time being. For housekeeping purposes we are closing stale PRs. If you'd still like to merge this code in, please open a new PR that has been merged and rebased with the `master` branch.

**Pull request closed without merging**